init/zsh-functions/_module

```diff
@@ -5,16 +5,31 @@
 # Copyright (C) 2017 Xavier Delaruelle <xavier.delaruelle@cea.fr>
 #
 
-_module_avail() {
-   module avail -t 2>&1 | sed '
-      /^-\+/d; /^\s*$/d;
-      /->.*$/d;
-      /:$/d;
-      /:ERROR:/d;
-      s#^\(.*\)/\(.\+\)(.*default.*)#\1\n\1\/\2#;
-      s#(.*)$##g;
-      s#\s*$##g;
-      s#/*$##g;'
+function _module_completion () {
+   emulate -L zsh
+   local _module_path_prefix
+   local -a _module_search_path _module_type_d _module_type_f
+   _module_path_prefix=$(echo ${(q)words[$CURRENT]} | sed -ne 's#\(.*/\).*#\1#p')
+   _module_search_paths=(${^${(@s/:/)MODULEPATH}}/$_module_path_prefix)
+
+   _module_type_d=()
+   _module_type_f=()
+   eval $(find $_module_search_paths -mindepth 1 -maxdepth 1 -not -name '.modulerc' -not -name '.version' -printf "_module_type_%Y+=${(q)_module_path_prefix}%P\n" 2> /dev/null)
```

Comment:
Using ``find`` rather ``module avail`` will miss the module aliases or the virtual modules. It should stay with ``module avail``.

Reply:
First point (details):
- assume you have a lot of different versions of  one module, then the current completion will list all version, which prevents fast-jumping to next module via <TAB> key.
- a second point is that we have a lot of module files distribute in our cluster and `module avail` take around 20 seconds (due to NFS) - not handy for a completion.
- furthermore the current completion version does not allow to load the default module e.g. `module load gcc` can't be achieved via zsh completion.

Second point (`find`):
-> The `find` command has the advantage that is just scan's for the direct available modules (`- maxdepth 1`) and it's fast + portable.

The best solution (from my point of view) would be to add a new switch to module avail
e.g. `module avail -s <argument>` which just list the direct access-able modules/directories (without descending). But this break's compability with the old-modules version (this is main-reason for the `find` command and virtual modules are quite new, right ?)
Another solution would be to cache everthing into a shell variable on startup - with the disadvantages, that it will miss updates.

In the end, it's just a suggestion - I enjoy a lot the completion the proposed way - did you tried it?
How ever it, will need some more work and this is the request to find out if you even interested in this kind of  enhancement.

Reply:
Thanks a lot for this detailed explanation. I get it now :)

There are 2 points to address:
1. ``avail`` may be slow with lots of modules and busy filesystem
2. default module names do not appear in completion results

For point ``1``, a cache feature will be introduced later this year to reduce as much as possible I/O operations. The new [virtual modules feature](https://modules.readthedocs.io/en/stable/MIGRATING.html#virtual-modules) introduced in v4.1 may also help if all modulefiles are converted into virtual modules empowered with some kind of a central registry of available software.

There is clearly an issue currently for point ``2`` as these default names should appear.

It seems also a very good idea to be able to only get the direct/simplified results of an ``avail``, especially in the completion use-case where at a given time you only need to see the names of the current searching level.

``find`` may work in your case but sometimes a modulepath contains directories with no module-related content and this should be filtered out from the results. There are also the virtual modules and aliases use-cases mentioned before. All of that to say that ``find`` does not fit all the supported module use-cases.

So the ``module avail -s <argument>`` as you say should be the best way and I will add some enhancement ticket on it.

This enhancement may also improve performances a bit, but not drastically as if you want to see level 1, it should go down level 2 to check there is at least one module.